robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java

```diff
@@ -684,14 +690,40 @@ private String createTempDir(String name) {
 
   // TODO move/replace this with packageManager
   @VisibleForTesting
-  static void registerBroadcastReceivers(Application application, AndroidManifest androidManifest) {
+  static void registerBroadcastReceivers(
+      Application application, AndroidManifest androidManifest, LoadedApk loadedApk) {
     for (BroadcastReceiverData receiver : androidManifest.getBroadcastReceivers()) {
       IntentFilter filter = new IntentFilter();
+      Intent intent = new Intent();
       for (String action : receiver.getActions()) {
         filter.addAction(action);
+        intent.setAction(action);
       }
       String receiverClassName = replaceLastDotWith$IfInnerStaticClass(receiver.getName());
-      application.registerReceiver((BroadcastReceiver) newInstanceOf(receiverClassName), filter);
+      BroadcastReceiver broadcastReceiver;
+      if (loadedApk != null && Build.VERSION.SDK_INT >= P) {
+        ClassLoader classLoader;
+        if (application.getBaseContext() != null) {
+          classLoader = application.getBaseContext().getClassLoader();
+        } else {
+          classLoader = loadedApk.getClassLoader();
+        }
+        // TODO utzcoz How about one receiver has multiple actions?
+        try {
+          broadcastReceiver =
```

Comment:
@hoisie IMO, it's not great enough now, and it needs feedback before merging.

Reply:
I resolved it with trying to initialize the receiver with all actions until initializing it succeed.